src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs

```diff
@@ -62,7 +62,13 @@ public static bool HasStarted(HttpContext context, out Activity? aspNetActivity)
     {
         PropagationContext propagationContext = textMapPropagator.Extract(default, context.Request, HttpRequestHeaderValuesGetter);
 
-        Activity? activity = AspNetSource.StartActivity(TelemetryHttpModule.AspNetActivityName, ActivityKind.Server, propagationContext.ActivityContext);
+        IEnumerable<KeyValuePair<string, object?>>? enumerable = null;
+        if (context.Request?.Unvalidated?.Path is string path)
+        {
+            enumerable = [new KeyValuePair<string, object?>("url.path", path)];
```

Comment:
instrumentations are generally not doing this due to concern over heap allocation. This must, however, be solved, and need to work with the owners (ActivitySource/Activity i.e runtime team) to see if they can support `TagList` or `StartActivity` overloads for taking 1,2,3 tags, so this allocation can be avoided.

Reply:
I couldn't find a good way to deal with the allocation The api requires IEnumerable so whatever happens it either has be a class or it'll be boxed. I tried TagList but decided the boxing of such a large struct was going to be more expensive than  what the compiler is likely to do here, which is probably a new single length array. I considered creating a class which can contain the single element and which is enumerable but it felt like overkill for a first contribution to the project.

Reply:
it is not possible to avoid this, without changes in the API offered by runtime itself.

The only instrumentation in the repo providing tags at startup is the sqlclient one, but only does so for fully static tag: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlActivitySourceHelper.cs#L25

Reply:
> instrumentations are generally not doing this due to concern over heap allocation. This must, however, be solved,

@Wraith2  Just to make it clear - I didn't mean that this should be solved in this PR or you should work with runtime team to solve this. Just stated that this is a problem that must be solved. (I can create a tracking issue in runtime and otel repos, once i find the old discussion/probably issue itself asking the same)

OTel convention requires a lot of tags to be provided at startup time itself, and it is not possible to do it without allocation in Activity, so this is a problem that needs to be solved even outside of this issue/pr.

Reply:
Couldn't find a tracking issue, but https://github.com/dotnet/runtime/issues/42784#issuecomment-733705736 has explained the problem. If you are up for it, please create a new issue in the runtime repo. Else, I can create one.

(Also, it is very unlikely that runtime will be able to make any changes for .NET 9 timeframe, so we are looking at .NET 10 timeframe only at this point)

Reply:
New proposal for a CreateActivity overload that takes a single kvp, https://github.com/dotnet/runtime/issues/103245

Reply:
Could we use a `ThreadStatic` for the storage?

```csharp
    [ThreadStatic]
    private static KeyValuePair<string, object?>[]? samplerTagStorage;

        IEnumerable<KeyValuePair<string, object?>>? tags;
        if (context.Request?.Unvalidated?.Path is string path)
        {
            var samplerTags = samplerTagStorage ??= new KeyValuePair<string, object?>[1];
            samplerTags[0] = new KeyValuePair<string, object?>("url.path", path);
            tags = samplerTags;
        }
        else
        {
            tags = null;
        }

        Activity? activity = AspNetSource.StartActivity(TelemetryHttpModule.AspNetActivityName, ActivityKind.Server, propagationContext.ActivityContext, tags);
```

Pretty sure sampling happens on the same thread so it is mostly safe. Some user could make a custom sampler that attempts to hold onto the tags but why would they?

Reply:
The tags will get added to the TagList in the activity and since KeyValuePair is a struct they're copied and the reference to the incoming array is not retained. So it'll be safe as long as no await is added along the current path.

I've added this change.